### PR TITLE
Fixed pull against docker store

### LINKED ISSUE #8138: Docker pull on docker store fails
Docker pull on images on docker store fails.  For instance,

```
docker pull store/ibmcorp/db2wh_ce:v2.9.0-db2wh_client-linux
```

Internal bug 2152747.

### PR DESCRIPTION
**Pull request by sflxn**
We changed the default docker registry URL back when we were
adding content trust.  That URL is causing this failure.  The
URL has been changed back to the same value that is hardcoded
in the docker distribution code.

Also fixed an error interpretation in the fetcher code to match
Docker's interpretation.  We were returning image not found
when the actual error was access denied.  This led the debugging
of this issue down the wrong path and will eventually lead us
down the wrong path with future bugs.

```
[full ci]
[parallel jobs=1]
```

Fixes #8138

### CODE REVIEW

File: lib/imagec/imagec_test.go [hunk 1]
```diff
@@ -601,8 +601,8 @@ func TestFetchScenarios(t *testing.T) {
 	// valid token but image is missing we shouldn't retry
 	_, _, err = FetchImageManifest(ctx, ic.Options, 1, ic.progressOutput)
 	if err != nil {
-		// we should get a ImageNotFoundError
-		if _, imageErr := err.(urlfetcher.ImageNotFoundError); !imageErr {
+		// we should get a AccessDenied
```
Comment (hickeng) [inline lib/imagec/imagec_test.go]: Will we ever get `ImageNotFoundErr`? The comment above notes that we have a valid token but a missing image. It surprises me that we therefore expect `AccessDenied`. I think the comment needs updating or the intent of the test needs updating.

File: tests/test-cases/Group1-Docker-Commands/1-02-Docker-Pull.md [hunk 2]
```diff
@@ -37,6 +37,9 @@ This test requires that an vSphere server is running and available.
     * gcr.io/google_samples/gb-redisslave:v1
     * gcr.io/google_samples/cassandra:v11
     * gcr.io/google_samples/cassandra:v12
+20. Verify image manifest digest against vanilla docker
+21. Attempt docker pull mitm
+22. Issue docker pull for store/ibmcorp/mqadvanced-server-dev:9.0.5.0
```
Comment (hickeng) [inline tests/test-cases/Group1-Docker-Commands/1-02-Docker-Pull.md]: 👍

File: tests/test-cases/Group1-Docker-Commands/1-02-Docker-Pull.robot [hunk 3]
```diff
@@ -136,7 +136,7 @@ Pull non-existent image
     ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} pull fakebadimage
     Log  ${output}
     Should Be Equal As Integers  ${rc}  1
-    Should contain  ${output}  image library/fakebadimage not found
+    Should contain  ${output}  pull access denied
```
Comment (hickeng) [inline tests/test-cases/Group1-Docker-Commands/1-02-Docker-Pull.robot]: Same question again - I'd expect either TagNotFound or ImageNotFound errors for an image that straight up doesn't exist. What I get from trying my local docker is a message that expresses either/or:
```
$ docker pull fakebadimage
Using default tag: latest
Error response from daemon: repository fakebadimage not found: does not exist or no pull access
```

File: lib/imagec/imagec.go [hunk 4]
```diff
@@ -138,7 +138,7 @@ var (
 
 const (
 	// DefaultDockerURL holds the URL of Docker registry
-	DefaultDockerURL = "registry.hub.docker.com"
+	DefaultDockerURL = "registry-1.docker.io"
```
Comment (hickeng) [inline lib/imagec/imagec.go]: This honestly seems like something that should be configurable so we can change the default registry. Doing so would also allow adjustment of this url if/when we find other impacts. Not needed for this PR but interesting in knowing viability.

Comment (zjs) [inline lib/imagec/imagec.go]: Really minor, but: given that we already have the docker code vendored, should we consider setting this value equal to the constant from their codebase?

https://github.com/vmware/vic/blob/24af8e8e5536ca0bcf7ee2a9e659822585654b4a/vendor/github.com/docker/docker/registry/config.go#L50-L53

It seems like that makes the relationship between this code and the docker client code a bit more obvious, hopefully reducing the chances that this constant is incorrectly changed in the future.

File: tests/test-cases/Group1-Docker-Commands/1-04-Docker-Create.robot [hunk 5]
```diff
@@ -113,7 +113,7 @@ Create simple top example
 Create fakeimage image
     ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} create fakeimage
     Should Be Equal As Integers  ${rc}  1
-    Should Contain  ${output}  Error: image library/fakeimage not found
+    Should Contain  ${output}  pull access denied
```
[the remark this answers is not in the document]
Comment (sflxn) [reply tests/test-cases/Group1-Docker-Commands/1-04-Docker-Create.robot]: 'not found' is not part of the message that the version of docker (1.13+ and current) displays.  I'm not sure where 'not found' came from.  I think we probably just created that message and checked for that message in the test.

The access denied is just a can message from docker.  It has nothing to do with whether it's authorized or not.

```
pull access denied for fakeimage, repository does not exist or may require 'docker login'
```

File: lib/imagec/imagec_test.go (hunk 1 above)
Comment (sflxn) [reply lib/imagec/imagec_test.go]: I have no idea where this ImageNotFoundErr came from.  I hesitate to remove something that might get displayed under a specific circumstance.  I can scan the vendored docker code to see if it's even a real message.

File: tests/test-cases/Group1-Docker-Commands/1-04-Docker-Create.robot (hunk 5 above)
[the remark this answers is not in the document]
Comment (sflxn) [reply tests/test-cases/Group1-Docker-Commands/1-04-Docker-Create.robot]: I found the message 'image -- not found' in the vendored docker code.  It's in the V1 puller; however, now that most if not all images have been migrated to V2, we probably will never see it anymore.

File: lib/imagec/imagec.go (hunk 4 above)
Comment (sflxn) [reply lib/imagec/imagec.go]: registry-1.docker.io is hard coded in docker code.  It's not a configurable option, but yes, doing so would give imageC some flexibility.

File: lib/imagec/docker.go [hunk 6]
```diff
@@ -136,8 +138,8 @@ func LearnAuthURL(op trace.Operation, options Options) (*url.URL, error) {
 
 	// Private registry returned the manifest directly as auth option is optional.
 	// https://github.com/docker/distribution/blob/master/docs/configuration.md#auth
-	if err == nil && options.Registry != DefaultDockerURL && fetcher.IsStatusOK() {
-		op.Debugf("%s does not support OAuth", url)
+	if err == nil && options.Registry != registry.DefaultV2Registry.Host && fetcher.IsStatusOK() {
+		log.Debugf("%s does not support OAuth", url)
```
Comment (hickeng) [inline lib/imagec/docker.go]: This seems like it should be using `op` given it's available. Same for the other new log message below.

File: lib/imagec/docker.go [hunk 7]
```diff
@@ -353,6 +355,13 @@ func FetchImageManifest(op trace.Operation, options Options, schemaVersion int,
 
 	manifestFileName, err := fetcher.Fetch(op, url, &reqHeaders, true, progressOutput)
 	if err != nil {
+		log.Debugf("Failed to fetch manifest: %s", err.Error())
+
+		switch err.(type) {
+		case urlfetcher.AccessDenied:
+			return nil, "", fmt.Errorf("pull access denied for %s, repository does not exist or may require 'docker login'", options.Reference.Name())
```
Comment (hickeng) [inline lib/imagec/docker.go]: 👍

**Comment by stuclem** [comment]
@hickeng in which release does this need to be documented? 1.4.1 or 1.4.3?

**Pull request merged**